tachyon/math/base/semigroups.h

```diff
@@ -92,17 +108,14 @@ class MultiplicativeSemigroup {
       typename internal::MultiplicativeSemigroupTraits<G>::ReturnTy;
 
   // Multiplication: a * b
-  template <
-      typename G2,
-      std::enable_if_t<internal::SupportsMul<G, G2>::value ||
-                       internal::SupportsMulInPlace<G, G2>::value>* = nullptr>
+  template <typename G2,
+            std::enable_if_t<internal::SupportsMul<G, G2>::value>* = nullptr>
   constexpr auto operator*(const G2& other) const {
     if constexpr (internal::SupportsMul<G, G2>::value) {
       const G* g = static_cast<const G*>(this);
       return g->Mul(other);
     } else {
-      G g = *static_cast<const G*>(this);
-      return g.MulInPlace(other);
+      static_assert(base::AlwaysFalse<G>);
     }
```

Comment:
I think once `SupportsMulInPlace` is removed as an enable if condition, there's no reason to have this if/else statement (+ same for `Add`) 
```suggestion
    const G* g = static_cast<const G*>(this);
    return g->Mul(other);
```